core/tracing/journal.go

```diff
@@ -0,0 +1,259 @@
+}
+
+// reset clears the journal, after this operation the journal can be used anew.
+// It is semantically similar to calling 'NewJournal', but the underlying slices
+// can be reused.
+func (j *journal) reset() {
+	j.entries = j.entries[:0]
+	j.validRevisions = j.validRevisions[:0]
+	j.nextRevisionId = 0
+}
+
+// snapshot returns an identifier for the current revision of the state.
+func (j *journal) snapshot() int {
+	id := j.nextRevisionId
+	j.nextRevisionId++
+	j.validRevisions = append(j.validRevisions, revision{id, j.length()})
+	return id
+}
+
+// revertToSnapshot reverts all state changes made since the given revision.
+func (j *journal) revertToSnapshot(revid int, hooks *Hooks) {
+	// Find the snapshot in the stack of valid snapshots.
+	idx := sort.Search(len(j.validRevisions), func(i int) bool {
+		return j.validRevisions[i].id >= revid
+	})
+	if idx == len(j.validRevisions) || j.validRevisions[idx].id != revid {
+		panic(fmt.Errorf("revision id %v cannot be reverted", revid))
+	}
+	snapshot := j.validRevisions[idx].journalIndex
+
+	// Replay the journal to undo changes and remove invalidated snapshots
+	j.revert(hooks, snapshot)
+	j.validRevisions = j.validRevisions[:idx]
+}
+
+// revert undoes a batch of journaled modifications.
+func (j *journal) revert(hooks *Hooks, snapshot int) {
+	for i := len(j.entries) - 1; i >= snapshot; i-- {
+		// Undo the changes made by the operation
+		j.entries[i].revert(hooks)
+	}
+	j.entries = j.entries[:snapshot]
+}
+
+// length returns the current number of entries in the journal.
+func (j *journal) length() int {
+	return len(j.entries)
+}
+
+func (j *journal) OnTxEnd(receipt *types.Receipt, err error) {
+	j.reset()
+	if j.hooks.OnTxEnd != nil {
+		j.hooks.OnTxEnd(receipt, err)
+	}
+}
+
+func (j *journal) OnEnter(depth int, typ byte, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int) {
+	j.revIds = append(j.revIds, j.snapshot())
+	if typ == CREATE || typ == CREATE2 {
+		j.lastCreator = &from
+	}
+	if j.hooks.OnEnter != nil {
+		j.hooks.OnEnter(depth, typ, from, to, input, gas, value)
+	}
+}
+
+func (j *journal) OnExit(depth int, output []byte, gasUsed uint64, err error, reverted bool) {
+	if j.lastCreator != nil {
+		j.lastCreator = nil
+	}
+	revId := j.revIds[len(j.revIds)-1]
+	j.revIds = j.revIds[:len(j.revIds)-1]
+	if reverted {
+		j.revertToSnapshot(revId, j.hooks)
+	}
+	if j.hooks.OnExit != nil {
+		j.hooks.OnExit(depth, output, gasUsed, err, reverted)
+	}
+}
+
+func (j *journal) OnBalanceChange(addr common.Address, prev, new *big.Int, reason BalanceChangeReason) {
+	j.entries = append(j.entries, balanceChange{addr: addr, prev: prev, new: new})
+	if j.hooks.OnBalanceChange != nil {
+		j.hooks.OnBalanceChange(addr, prev, new, reason)
+	}
+}
+
+func (j *journal) OnNonceChange(addr common.Address, prev, new uint64) {
+	// When a contract is created, the nonce of the creator is incremented.
+	// This change is not reverted when the creation fails.
```

Comment:
Hm, what? Doesn't that depend on ... things.. ? Or is it always the case?

Reply:
For evm-creates (as opposed to create-tx where the flow is different), the new scope begins here: https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go#L425 , and the nonce-bump happens a few lines later: https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go#L442

However, the _actual_ snapshot is taken even further below: https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go#L479 . 

So that's the mismatch you see. What the tracing percieves is 
```
SCOPE_START{
    NONCE ++
    OTHER_STUFF
}
```
But the reality is 
```
NONCE++
SCOPE_START{
    OTHER_STUFF
}
```
So when the scope reverts, you see this weird inconsistency that somehow nonce is a special snowflake. 


The proper fix would be to move the `captureBegin`-invocation down, so it happens close to where we take the snapshot. The nonce-bump belongs to the parent scope.